sdk/dotnet/Docker.cs

```diff
@@ -0,0 +1,644 @@
+        /// </summary>
+        /// <param name="cmd">Command name to run.</param>
+        /// <param name="args">An array of command arguments.</param>
+        /// <param name="logResource">Used to specify the resource to associate command output with. Stderr messages
+        /// are always sent (since they may contain important information about something that's gone wrong).
+        /// Stdout messages will be logged ephemerally to this resource.  This lets the user know there is
+        /// progress, without having that dumped on them at the end.  If an error occurs though, the stdout
+        /// content will be printed.</param>
+        /// <param name="reportFullCommandLine"></param>
+        /// <param name="reportErrorAsWarning"></param>
+        /// <param name="stdin">If defined, its contents are piped into stdin for the child process.</param>
+        /// <param name="env"></param>
+        /// <returns></returns>
+        private static async Task<(int, string)> RunCommandThatCanFail(
+            string cmd,
+            string[] args,
+            Resource logResource,
+            bool reportFullCommandLine,
+            bool reportErrorAsWarning,
+            string? stdin = null,
+            Dictionary<string, string>? env = null)
+        {
+            // Let the user ephemerally know the command we're going to execute.
+            Log.Info($"Executing {GetCommandLineMessage(cmd, args, reportFullCommandLine, env)}", logResource, ephemeral: true);
+
+            // Generate a unique stream-ID that we'll associate all the docker output with. This will allow
+            // each spawned CLI command's output to associated with 'resource' and also streamed to the UI
+            // in pieces so that it can be displayed live.  The stream-ID is so that the UI knows these
+            // messages are all related and should be considered as one large message (just one that was
+            // sent over in chunks).
+            //
+            // We use Math.random here in case our package is loaded multiple times in memory (i.e. because
+            // different downstream dependencies depend on different versions of us).  By being random we
+            // effectively make it completely unlikely that any two cli outputs could map to the same stream
+            // id.
+            //
+            // Pick a reasonably distributed number between 0 and 2^30.  This will fit as an int32
+            // which the grpc layer needs.
+            var streamID = (int)Math.Floor(new Random().NextDouble() * (1 << 30));
+
+            var process = new Process();
+            process.StartInfo.FileName = cmd;
+            process.StartInfo.Arguments = Utils.EscapeArguments(args);
+            process.StartInfo.RedirectStandardInput = stdin != null;
+            process.StartInfo.RedirectStandardOutput = true;
+            process.StartInfo.RedirectStandardError = true;
+            process.Start();
```

Comment:
we might want to consider a try/catch around this code.  .net process stuff can be finicky in my experience

Reply:
Added a `try/catch` around the process call, returning `(1, ex.ToString())` in case an exception is thrown. Is that what you intended?